[DOCS] CCR disaster recovery (#91491)
elastic/elasticsearch: docs/reference/ccr/bi-directional-disaster-recovery.asciidoc

[role="xpack"]
[[ccr-disaster-recovery-bi-directional-tutorial]]
=== Tutorial: Disaster recovery based on bi-directional {ccr}
++++
<titleabbrev>Bi-directional disaster recovery</titleabbrev>
++++

Learn how to set up disaster recovery between two clusters based on

> bi-directional {ccr}. The following tutorial is designed for immutable indices
> (data no longer need to be updated after writing) like logs, metrics, observability, 
> and security events. If you have mutable indices, you should follow <<disaster-recovery,the uni-directional cross-cluster replication disaster recovery tutorial>>.

Author: Turns out data stream supports `update_by_query` and `delete_by_query`. https://www.elastic.co/guide/en/elasticsearch/reference/master/use-a-data-stream.html#update-docs-in-a-data-stream-by-query. As such, I will remove the "immutable" restriction. 
 
```suggestion
bi-directional {ccr}. The following tutorial is designed for data streams which supports <<{ref}/use-a-data-stream.html#update-docs-in-a-data-stream-by-query,update_by_query>> and <<{ref}/use-a-data-stream.html#delete-docs-in-a-data-stream-by-query,delete_by_query>>. You can only perform these actions on the leader index. 
```

This tutorial works with Logstash as the source of ingestion. It takes
advantage of a logstash feature where <<{logstash-ref}/plugins-outputs-elasticsearch,the output can be load balanced
across an array of hosts specified>>. Beats and agents currently do not 
support multiple outputs. It should also be possible to set up a proxy 
(Load Balancer) to redirect traffic without Logstash in this tutorial. 

* Setting up a remote cluster on `clusterA` and `clusterB`.
* Setting up bi-directional cross-cluster replication with exclusion patterns.
* Setting up Logstash with multiple hosts to allow automatic load balancing and switching during disasters.

image::images/ccr-bi-directional-disaster-recovery.png[Bi-directional cross cluster replication failover and failback]

==== Initial setup
. Set up a remote cluster on both clusters.
+
[source,console]
----
### On cluster A ###
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "clusterB": {
          "mode": "proxy",
          "skip_unavailable": true,
          "server_name": "clusterb.es.australia-southeast1.gcp.elastic-cloud.com",
          "proxy_socket_connections": 18,

>           "proxy_address": "clusterb.es.australia-southeast1.gcp.elastic-cloud.com:9400",

Author: ```suggestion
          "proxy_address": "clusterb.es.australia-southeast1.gcp.elastic-cloud.com:9400"
```

        }
      }
    }
  }
}
### On cluster B ###
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "clusterA": {
          "mode": "proxy",
          "skip_unavailable": true,
          "server_name": "clustera.es.australia-southeast1.gcp.elastic-cloud.com",
          "proxy_socket_connections": 18,

>           "proxy_address": "clustera.es.australia-southeast1.gcp.elastic-cloud.com:9400",

Author: ```suggestion
          "proxy_address": "clustera.es.australia-southeast1.gcp.elastic-cloud.com:9400"
```

** 50 documents in logs-generic-default-replicated_from_clustera
** 150 documents in logs-generic-default

. If you have logstash running at this time, you will also observe traffic is
sending to both clusters.

==== Perform update or delete by query
It is possible to update or delete the documents but you can only perform these actions on the leader index.
. First identify which backing index contains the document you want to update.
+
[source,console]
----
### On either of the cluster ###
GET logs-generic-default*/_search?filter_path=hits.hits._index
{
"query": {
    "match": {
      "event.sequence": "97"
    }
  }
}
----
+
* If the hits returns ` "_index": ".ds-logs-generic-default-replicated_from_clustera-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on `cluster A`.
* If the hits returns ` "_index": ".ds-logs-generic-default-replicated_from_clusterb-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on `cluster B`.
* If the hits returns ` "_index": ".ds-logs-generic-default-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on the same cluster where you perform the search query.

. Perform the update (or delete) by query

> ### On the cluster identified from the previous step ###
> POST logs-generic-default/_update_by_query
> {
>   "query": {
>     "match": {
>       "event.sequence": "97"
>     }
>   },
>   "script": {
>     "source": "ctx._source.event.original = params.new_event",
>     "lang": "painless",
>     "params": {
>       "new_event": "FOOBAR"
>     }
>   }
> }

Author: ```suggestion
+
[source,console]
----
### On the cluster identified from the previous step ###
POST logs-generic-default/_update_by_query
{
  "query": {
    "match": {
      "event.sequence": "97"
    }
  },
  "script": {
    "source": "ctx._source.event.original = params.new_event",
    "lang": "painless",
    "params": {
      "new_event": "FOOBAR"
    }
  }
}
----
```

[role="xpack"]
[[ccr-disaster-recovery-bi-directional-tutorial]]
=== Tutorial: Disaster recovery based on bi-directional {ccr}
++++
<titleabbrev>Bi-directional disaster recovery</titleabbrev>
++++

Learn how to set up disaster recovery between two clusters based on
bi-directional {ccr}. The following tutorial is designed for data streams which supports 

> <<{ref}/use-a-data-stream.html#update-docs-in-a-data-stream-by-query,update_by_query>> and
>  <<{ref}/use-a-data-stream.html#delete-docs-in-a-data-stream-by-query,delete_by_query>>. 

Author: ```suggestion
<<update-docs-in-a-data-stream-by-query,update_by_query>> and
 <<delete-docs-in-a-data-stream-by-query,delete_by_query>>. 
```

 You can only perform these actions on the leader index. 

This tutorial works with Logstash as the source of ingestion. It takes

> advantage of a logstash feature where <<{logstash-ref}/plugins-outputs-elasticsearch,the output can be load balanced

Author: ```suggestion
advantage of a logstash feature where {logstash-ref}/plugins-outputs-elasticsearch.html[logstash  output to Elasticsearch] can be load balanced
```

+
TIP: You should also redirect all search traffic to the `clusterB` cluster during this time. 

. Observe two data streams on `cluster B` now contain a different number of documents. 
+
* data streams On cluster A (Dead) 
** 50 documents in logs-generic-default-replicated_from_clusterb 
** 50 documents in logs-generic-default
* data streams On cluster B (Alive) 
** 50 documents in logs-generic-default-replicated_from_clustera
** 150 documents in logs-generic-default

==== Failback when `clusterA` comes back
. You can simulate this by turning `cluster A` back. 
. Observe data ingested to `cluster B` during `cluster A` 's downtime will be
automatically replicated. 
+
* data streams On cluster A
** 150 documents in logs-generic-default-replicated_from_clusterb 
** 50 documents in logs-generic-default
* data streams On cluster B
** 50 documents in logs-generic-default-replicated_from_clustera
** 150 documents in logs-generic-default

. If you have logstash running at this time, you will also observe traffic is
sending to both clusters.

==== Perform update or delete by query
It is possible to update or delete the documents but you can only perform these actions on the leader index.

> . First identify which backing index contains the document you want to update.

Author: It appears it needs a blankline before hand in order to create a list.
```suggestion

. First identify which backing index contains the document you want to update.
```

** 50 documents in logs-generic-default
* data streams On cluster B (Alive) 
** 50 documents in logs-generic-default-replicated_from_clustera
** 150 documents in logs-generic-default

==== Failback when `clusterA` comes back
. You can simulate this by turning `cluster A` back. 
. Observe data ingested to `cluster B` during `cluster A` 's downtime will be
automatically replicated. 
+
* data streams On cluster A
** 150 documents in logs-generic-default-replicated_from_clusterb 
** 50 documents in logs-generic-default
* data streams On cluster B
** 50 documents in logs-generic-default-replicated_from_clustera
** 150 documents in logs-generic-default

. If you have logstash running at this time, you will also observe traffic is
sending to both clusters.

==== Perform update or delete by query
It is possible to update or delete the documents but you can only perform these actions on the leader index.

. First identify which backing index contains the document you want to update.
+
[source,console]
----
### On either of the cluster ###
GET logs-generic-default*/_search?filter_path=hits.hits._index
{
"query": {
    "match": {
      "event.sequence": "97"
    }
  }
}
----
+

> * If the hits returns ` "_index": ".ds-logs-generic-default-replicated_from_clustera-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on `cluster A`.
> * If the hits returns ` "_index": ".ds-logs-generic-default-replicated_from_clusterb-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on `cluster B`.
> * If the hits returns ` "_index": ".ds-logs-generic-default-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on the same cluster where you perform the search query.

Author: ```suggestion
* If the hits returns `"_index": ".ds-logs-generic-default-replicated_from_clustera-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on `cluster A`.
* If the hits returns `"_index": ".ds-logs-generic-default-replicated_from_clusterb-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on `cluster B`.
* If the hits returns `"_index": ".ds-logs-generic-default-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on the same cluster where you perform the search query.
```

[role="xpack"]
[[ccr-disaster-recovery-bi-directional-tutorial]]
=== Tutorial: Disaster recovery based on bi-directional {ccr}
++++
<titleabbrev>Bi-directional disaster recovery</titleabbrev>
++++

////
[source,console]
----

> PUT logs-generic-default

Author: ```suggestion
PUT _data_stream/logs-generic-default
```

* Setting up Logstash with multiple hosts to allow automatic load balancing and switching during disasters.

image::images/ccr-bi-directional-disaster-recovery.png[Bi-directional cross cluster replication failover and failback]

==== Initial setup
. Set up a remote cluster on both clusters.
+
[source,console]
----
### On cluster A ###
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "clusterB": {
          "mode": "proxy",
          "skip_unavailable": true,
          "server_name": "clusterb.es.australia-southeast1.gcp.elastic-cloud.com",
          "proxy_socket_connections": 18,
          "proxy_address": "clusterb.es.australia-southeast1.gcp.elastic-cloud.com:9400"
        }
      }
    }
  }
}
### On cluster B ###
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "clusterA": {
          "mode": "proxy",
          "skip_unavailable": true,
          "server_name": "clustera.es.australia-southeast1.gcp.elastic-cloud.com",
          "proxy_socket_connections": 18,
          "proxy_address": "clustera.es.australia-southeast1.gcp.elastic-cloud.com:9400"
        }
      }
    }
  }
}

> ----

Author: ```suggestion
----
// TEST[setup:host]
// TEST[s/"server_name": "clustera.es.australia-southeast1.gcp.elastic-cloud.com",//]
// TEST[s/"server_name": "clusterb.es.australia-southeast1.gcp.elastic-cloud.com",//]
// TEST[s/"proxy_socket_connections": 18,//]
// TEST[s/clustera.es.australia-southeast1.gcp.elastic-cloud.com:9400/\${transport_host}/]
// TEST[s/clusterb.es.australia-southeast1.gcp.elastic-cloud.com:9400/\${transport_host}/]
```

. Set up bi-directional cross-cluster replication.
+
[source,console]
----
### On cluster A ###
PUT /_ccr/auto_follow/logs-generic-default
{
  "remote_cluster": "clusterB",
  "leader_index_patterns": [
    ".ds-logs-generic-default-20*"
  ],
  "leader_index_exclusion_patterns":"{{leader_index}}-replicated_from_clustera",
  "follow_index_pattern": "{{leader_index}}-replicated_from_clusterb"
}

### On cluster B ###
PUT /_ccr/auto_follow/logs-generic-default
{
  "remote_cluster": "clusterA",
  "leader_index_patterns": [
    ".ds-logs-generic-default-20*"
  ],
  "leader_index_exclusion_patterns":"{{leader_index}}-replicated_from_clusterb",
  "follow_index_pattern": "{{leader_index}}-replicated_from_clustera"
}

> ----

Author: ```suggestion
----
// TEST[setup:remote_cluster]
// TEST[s/clusterA/remote_cluster/]
// TEST[s/clusterB/remote_cluster/]
```

This tutorial works with {ls} as the source of ingestion. It takes advantage of a {ls} feature where {logstash-ref}/plugins-outputs-elasticsearch.html[the {ls} output to {es}] can be load balanced across an array of hosts specified. {beats} and {agents} currently do not 
support multiple outputs. It should also be possible to set up a proxy 
(load balancer) to redirect traffic without {ls} in this tutorial. 

* Setting up a remote cluster on `clusterA` and `clusterB`.
* Setting up bi-directional cross-cluster replication with exclusion patterns.
* Setting up {ls} with multiple hosts to allow automatic load balancing and switching during disasters.

image::images/ccr-bi-directional-disaster-recovery.png[Bi-directional cross cluster replication failover and failback]

[[ccr-tutorial-initial-setup]]
==== Initial setup
. Set up a remote cluster on both clusters.
+
[source,console]
----
### On cluster A ###
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "clusterB": {
          "mode": "proxy",
          "skip_unavailable": true,

>           "server_name": "clusterb.es.australia-southeast1.gcp.elastic-cloud.com",
>           "proxy_socket_connections": 18,
>           "proxy_address": "clusterb.es.australia-southeast1.gcp.elastic-cloud.com:9400"

Author: ```suggestion
          "server_name": "clusterb.es.region-b.gcp.elastic-cloud.com",
          "proxy_socket_connections": 18,
          "proxy_address": "clusterb.es.region-b.gcp.elastic-cloud.com:9400"
```

        }
      }
    }
  }
}
### On cluster B ###
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "clusterA": {
          "mode": "proxy",
          "skip_unavailable": true,

>           "server_name": "clustera.es.australia-southeast1.gcp.elastic-cloud.com",
>           "proxy_socket_connections": 18,
>           "proxy_address": "clustera.es.australia-southeast1.gcp.elastic-cloud.com:9400"

Author: ```suggestion
          "server_name": "clustera.es.region-a.gcp.elastic-cloud.com",
          "proxy_socket_connections": 18,
          "proxy_address": "clustera.es.region-a.gcp.elastic-cloud.com:9400"
```

        }
      }
    }
  }
}
----
// TEST[setup:host]

> // TEST[s/"server_name": "clustera.es.australia-southeast1.gcp.elastic-cloud.com",//]
> // TEST[s/"server_name": "clusterb.es.australia-southeast1.gcp.elastic-cloud.com",//]
> // TEST[s/"proxy_socket_connections": 18,//]
> // TEST[s/clustera.es.australia-southeast1.gcp.elastic-cloud.com:9400/\${transport_host}/]
> // TEST[s/clusterb.es.australia-southeast1.gcp.elastic-cloud.com:9400/\${transport_host}/]

Author: ```suggestion
// TEST[s/"server_name": "clustera.es.region-a.gcp.elastic-cloud.com",//]
// TEST[s/"server_name": "clusterb.es.region-b.gcp.elastic-cloud.com",//]
// TEST[s/"proxy_socket_connections": 18,//]
// TEST[s/clustera.es.region-a.gcp.elastic-cloud.com:9400/\${transport_host}/]
// TEST[s/clusterb.es.region-b.gcp.elastic-cloud.com:9400/\${transport_host}/]
```

  "follow_index_pattern": "{{leader_index}}-replicated_from_clusterb"
}

### On cluster B ###
PUT /_ccr/auto_follow/logs-generic-default
{
  "remote_cluster": "clusterA",
  "leader_index_patterns": [
    ".ds-logs-generic-default-20*"
  ],
  "leader_index_exclusion_patterns":"{{leader_index}}-replicated_from_clusterb",
  "follow_index_pattern": "{{leader_index}}-replicated_from_clustera"
}
----
// TEST[setup:remote_cluster]
// TEST[s/clusterA/remote_cluster/]
// TEST[s/clusterB/remote_cluster/]
+
IMPORTANT: Existing data on the cluster will not be replicated by
`_ccr/auto_follow` even though the patterns may match. This function will only
replicate newly created backing indices (as part of the data stream).
+
IMPORTANT: Use `leader_index_exclusion_patterns` to avoid recursion.
+
TIP: `follow_index_pattern` allows lowercase characters only.
+
TIP: This step cannot be executed via the {kib} UI due to the lack of an exclusion
pattern in the UI. Use the API in this step.

. Set up the {ls} configuration file.
+
This example uses the input generator to demonstrate the document
count in the clusters. Reconfigure this section
to suit your own use case. 
+
[source,logstash]
----

> ### On Logstash server ###

Reviewer: Would logstash not be a single point of failure?

I think the original blog post would index data locally in the current DC to the local cluster, avoiding this issue to some degree.

Author: Exactly. Unfortunately, the original blog post having DC locality isolation does not satisfy most of the bi-directional CCR use cases where users would have a single source of ingestion going into two separated cluster. That's part of the reason I was working on this tutorial to highlight this use case.

There are other possibility avoiding Logstash being the single point of failure, but it is outside the scope of this tutorial.

### This is a logstash config file ###
input {
  generator{
    message => 'Hello World'
    count => 100
  }
}
output {
  elasticsearch {

>     hosts => ["https://clustera.es.australia-southeast1.gcp.elastic-cloud.com:9243","https://clusterb.es.australia-southeast1.gcp.elastic-cloud.com:9243"]

Reviewer: The hosts here are int he the same region. I wonder if that is the example we want to target here? The use case is not clear to me.

Author: ```suggestion
    hosts => ["https://clustera.es.region-a.gcp.elastic-cloud.com:9243","https://clusterb.es.region-b.gcp.elastic-cloud.com:9243"]
```

Good point. I've change them to `region-a` and `region-b`. Will change other parts in the tutorial to reflect this too.

** 150 documents in `logs-generic-default`

. If you have {ls} running at this time, you will also observe traffic is
sent to both clusters.

==== Perform update or delete by query
It is possible to update or delete the documents but you can only perform these actions on the leader index.

. First identify which backing index contains the document you want to update.
+
[source,console]
----
### On either of the cluster ###
GET logs-generic-default*/_search?filter_path=hits.hits._index
{
"query": {
    "match": {
      "event.sequence": "97"
    }
  }
}
----
+
* If the hits returns `"_index": ".ds-logs-generic-default-replicated_from_clustera-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on `cluster A`.
* If the hits returns `"_index": ".ds-logs-generic-default-replicated_from_clusterb-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on `cluster B`.
* If the hits returns `"_index": ".ds-logs-generic-default-<yyyy.MM.dd>-*"`, then you need to proceed to the next step on the same cluster where you performed the search query.

. Perform the update (or delete) by query:
+
[source,console]
----
### On the cluster identified from the previous step ###

> POST logs-generic-default/_update_by_query

Reviewer: Would they not simply use individual updates or deletes instead (maybe through bulk)? I can imagine the query above giving out documents that are on both sides and they'd have to partition the document `_id` to use on each cluster.

Author: Another common request from users is to perform update or deletes in bi-directional CCR.

The purpose of this section is to demonstrate the possibility of doing so. Users are still welcome to bulk delete/update on the leader index as they wish.

{
  "query": {
    "match": {
      "event.sequence": "97"
    }
  },
  "script": {
    "source": "ctx._source.event.original = params.new_event",
    "lang": "painless",
    "params": {
      "new_event": "FOOBAR"
    }
  }
}

> ----

Author: ```suggestion
----
+ TIP: If a soft delete is merged away before it can be replicated to a follower the following process will fail due to incomplete history on the leader, see <<ccr-index-soft-deletes-retention-period, index.soft_deletes.retention_lease.period>> for more details.
```
